File: sort/bubble_sort.go
package main

import (
    "fmt"
)

/**
 * 冒泡排序
 * Time Complexity O(n^2)
 * Space Complexity 总共O(n)，辅助空间O(1)
 */
func BubbleSort(arr []int){
    length := len(arr)

    for i:=0; i<length-1; i++ {
        isChange := false
        for j:=0; j<length-1-i; j++ {
            if arr[j] > arr[j+1] {
                arr[j], arr[j+1] = arr[j+1], arr[j]
                isChange = true
            }
        }
        if isChange == false {
            break
        }
    }
}

func main() {
    arr := []int{22, 34, 3, 32, 82, 55, 89, 50, 37, 5, 64, 35, 9, -1, 70}
    fmt.Println(arr)
    BubbleSort(arr)
    fmt.Println(arr)
}

File: sort/BubbleSort.java
import java.util.Arrays;

class BubbleSort {


    public static void main(String[] args) {
        int[] arr = new int[]{5,2,18,3,11,90,45,22};
        arr = bubbleSort(arr);
        System.out.println(Arrays.toString(arr));
    }


    /**
     * Time Complexity O(n^2)
     * Space Complexity O(1)
     */
    public static int[] bubbleSort(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return arr;
        }

        for (int i = 0; i < arr.length - 1; i++) {
            boolean change = false;
            for (int j = 0; j < arr.length - i - 1; j++) {
                if (arr[j] > arr[j+1]) {
                    int tmp = arr[j+1];
                    arr[j+1] = arr[j];
                    arr[j] = tmp;
                    change = true;
                }
            }
            if (!change) {
                break;
            }
        }
        return arr;
    }


}
